Prepare for v2

HeahDude opened the pull request:
- Require PHP 5.5 or 7.0 and Symfony 3.4 minimum
- Refactored extension and enabled autowiring
- [BC Break] Removed classes parameters
- [BC Break] Removed the form data transformer
- added a new text form type extension with a purifier listener to purify submitted data in all text based fields, using opt-in and custom profile thanks to dedicated options
- added a new "exercise.html_purifier" tag to make custom purifier implementations available as profile through form options and Twig filter
- added a purifiers registry to lazy load purifiers everywhere
- added a Twig HTMLPurifierRuntime for better performances
- upgraded the LICENSE and README files

TODO
- [x] rebase after #47 has been merged

_____
For other PRs:
- [ ] merge https://github.com/heahprod/HTMLPurifierBundle/pull/2
- [ ] move classes in a `src/` folder and tests in `tests/`

HeahDude commented:
And btw if you're still interested in maintaining this bundle and any feature in there I would be glad to open a proper PR and add a BC layer if needed.

spolischook commented:
@HeahDude we need to made some branch for that. Please make PR to 4.0 branch

HeahDude commented:
:) So what do you need? I just remove the file headers and submit it?

spolischook commented:
Yep

spolischook commented:
Please check the tests locally after update - it must be green

HeahDude commented:
Should be good now.

spolischook commented:
@HeahDude many thanks for your contribution, I'll review you PR and let you know about results

HeahDude commented:
With pleasure :)

I'm not sure about the Symphony 2 compatibility, I have no project to test that right now.

spolischook commented:
It should be tested with symfony from composer.json

HeahDude commented:
I confirm tests pass when forcing Symphony 2 requirements.

HeahDude commented:
I fixed a few things about class name convention, some typos here and there, and enforced lazy loading while keeping BC with Symfony 2, that I had left around.

A review would be very welcome :). Thanks!

spolischook commented:
Please add phpunit as dev dependency to composer.json, and check that tests is passed

HeahDude commented:
This is the case.

HeahDude commented:
I've removed a form option about dynamic configs that do not play well with cache.

I've opened https://github.com/heahprod/HTMLPurifierBundle/pull/2 separately to ease the review but it should be merge before release as it breaks BC. It fixes #22 with custom profiles and #26.

Support for Symfony 2 has been drop to ease the code and rely on powerful features. So we should merge #44 to keep BC and full support of 0.2.x versions with Symfony 3.4 and 4.

I've updated the description.

nicolas-grekas commented on DependencyInjection/Compiler/HTMLPurifierPass.php:
```diff
@@ -0,0 +1,44 @@
+        $locator = $container->getDefinition('exercise_html_purifier.purifiers_locator');
+
+        foreach ($container->findTaggedServiceIds('exercise.html_purifier') as $id => $tags) {
+            foreach ($tags as $tag) {
```
replace foreach by $tags[0]

nicolas-grekas commented on Form/TypeExtension/HTMLPurifierTextTypeExtension.php:
```diff
@@ -0,0 +1,71 @@
+    private $cachedListeners;
+    private $purifiersRegistry;
+
+    public function __construct(HTMLPurifiersRegistry $registry)
```
the type hint should be against an interface

nicolas-grekas commented on DependencyInjection/ExerciseHTMLPurifierExtension.php:
```diff
@@ -25,92 +27,61 @@ public function load(array $configs, ContainerBuilder $container)
+
+        $purifiers = [];
+        $purifiersLocator = $container
+            ->register('exercise_html_purifier.purifiers_locator', ServiceLocator::class)
```
you should use ServiceLocatorTagPass::register() instead (with 3rd arg)

nicolas-grekas commented on Form/Listener/HTMLPurifierListener.php:
```diff
@@ -0,0 +1,38 @@
+{
+    private $purifier;
+
+    public function __construct(\HTMLPurifier $purifier)
```
this should be lazy: inject two args: a registry + $profile

spolischook commented:
Thanks @nicolas-grekas for review!

yakobe commented:
It this PR ready to be merged? It's one of the last things preventing a symfony 4 upgrade for us. Is there any way I can help out?

spolischook commented:
@yakobe can you try it manually and prove that it's works?

alister commented:
I've tested #44 with a Composer repository stanza (it works as expected, and the public service works), but I think that this PR does too much. If there was one PR to 'make this work with Symfony 4' (ie: only to make the service public), I'd take another look at it, and then I'd also look at other PRs for the additional functionality.

stof commented:
@alister see #47 for the Symfony 4 support.

stof commented on DependencyInjection/Compiler/HTMLPurifierPass.php:
```diff
@@ -0,0 +1,57 @@
+            $container->setAlias('exercise_html_purifier.'.$tags[0]['profile'], $id);
+        }
+
+        $locator->setArguments($purifiers);
```
altering a ServiceLocator service is a bad idea, as the ServiceLocatorPass performs de-duplication of the service locators based on their config. You should rather ask the  ServiceLocatorPass to register the service locator for the new set of arguments (or even better, configure the service locator *only* here in the compiler pass instead of doing it first in the DI extension and then here to find additional profiles)

HeahDude replied:
Fixed, thanks!

stof commented on Form/TypeExtension/HTMLPurifierTextTypeExtension.php:
```diff
@@ -0,0 +1,77 @@
+    private function getHTMLPurifierListener($profile)
+    {
+        if (isset($this->cachedListeners[$profile])) {
+            return $this->cachedListeners[$profile];
```
this caching is useless IMO. The instantiation of the listener is cheap, and such memoization can lead to memory leaks in long-running processes.

HeahDude replied:
Fixed.

stof commented on Form/Listener/HTMLPurifierListener.php:
```diff
@@ -0,0 +1,58 @@
+     */
+    private function getPurifier()
+    {
+        if (null !== $this->cachedPurifier) {
```
this memoization is not worth it IMO

HeahDude replied:
Fixed.

stof commented on README.md:
```diff
@@ -122,84 +76,89 @@ option to suppress the default path.
+    #...
+    
+    exercise_html_purifier.custom:
+        alias: \HTMLPurifier
```
that broken. You need to configure a `HTMLPurifier` alias targetting `exercise_html_purifier.custom`, not the opposite

HeahDude replied:
Fixed, plus added alternative syntax.

stof commented on DependencyInjection/ExerciseHTMLPurifierExtension.php:
```diff
@@ -25,92 +28,55 @@ public function load(array $configs, ContainerBuilder $container)
         }
 
-        $container->setParameter('exercise_html_purifier.cache_warmer.serializer.paths', array_unique($paths));
+        $container->register(HTMLPurifiersRegistry::class)
```
the registry should not be registered using its class name, but using an id instead. We don't want the `HTMLPurifiersRegistry` to be available for autowiring, only the interface.

HeahDude replied:
Fixed.

stof commented on Resources/config/html_purifier.xml:
```diff
@@ -2,26 +2,19 @@
-            <argument>%exercise_html_purifier.cache_warmer.serializer.paths%</argument>
-            <argument type="service" id="exercise_html_purifier.default" />
-            <tag name="kernel.cache_warmer" />
+        <defaults public="false" autowire="true" autoconfigure="true" />
```
For shared bundles, using explicit configuration (rather than autowiring and autoconfiguration) is better, as it ensures that the bundle maintainer knows how the bundle gets configured.

HeahDude replied:
Fixed.

stof commented:
> [BC Break] Removed the form data transformer

Why removing the data transformer ? This is a good way to perform HTML filtering.

XWB commented:
@HeahDude Any updates on this PR?

HeahDude commented:
@XWB updated :).

@stof Thank you very much for the review.

>Why removing the data transformer ? This is a good way to perform HTML filtering.

No... definitely not, the transformer is a wrong usage. Implementing the interface means changing the representation, one way to another. As long as an implementation cannot verify:
```php
// Given $a
$b = $transformer->transform($a);

$a === $transformer->reverseTransform($b); // must be true
```
then this is either wrong or a hack ([even if we hack Symfony](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/Extension/Core/Type/TextType.php#L56), though I'm very confident I'll get rid of that for 5.0).

A view transformer is called three times in a submission process by the way.

Here we just want to normalize the submitted data, we don't want to alter model or norm data in any way. Another reason is that now the purify option superseed the default trim processing.
*Bref*, it's much much cleaner like this IMO.

Anyone to test/review it? Thanks!

HeahDude commented:
https://github.com/heahprod/HTMLPurifierBundle/pull/2 has been rebased and fixed too.

XWB commented:
@stof Can you have a look as well?

stof commented on DependencyInjection/ExerciseHTMLPurifierExtension.php:
```diff
@@ -25,92 +27,55 @@ public function load(array $configs, ContainerBuilder $container)
+            ->setPublic(false)
+        ;
+        $container->setAlias(HTMLPurifiersRegistryInterface::class, 'exercise_html_purifier.purifiers_registry');
+        $container->setAlias(\HTMLPurifier::class, 'exercise_html_purifier.default');
```
autoloading aliases should always be private

HeahDude replied:
Fixed.

stof commented on HTMLPurifiersRegistry.php:
```diff
@@ -0,0 +1,35 @@
+            return $this->purifiersLocator->get($profile);
+        }
+
+        throw new \InvalidArgumentException(sprintf('The profile "%s" is not registered.', $profile));
```
if you let `$this->purifiersLocator->get` trigger an exception instead, it might suggest alternative names.

HeahDude replied:
Ha, didn't know that thanks. Fixed.

stof commented on README.md:
```diff
@@ -122,84 +76,92 @@ option to suppress the default path.
   [configuration documentation]: http://htmlpurifier.org/live/configdoc/plain.html
 
-## Cache Warming ##
+## Configuration in Symfony 4
```
isn't this exactly the same than for Symfony 3, except for the path to the YAML file, which is not actually related to 3 vs 4 but to Flex vs non-Flex (you can use Flex with 3.4)

HeahDude replied:
Right, fixed.

stof commented on README.md:
```diff
@@ -214,3 +176,42 @@ as follows:
+
+## Purifiers Registry
+
+A `Heah\HtmlPurifierBundle\HtmlPurifiersRegistry` class is registered by default
```
wrong namespace

HeahDude replied:
Good catch! Fixed.

stof commented on Twig/HTMLPurifierRuntime.php:
```diff
@@ -0,0 +1,43 @@
+
+class HTMLPurifierRuntime implements RuntimeExtensionInterface
+{
+    private $purifiersRegistry = [];
```
initializing it to an empty array is wrong. This variable is not meant to contain an array

HeahDude replied:
Artefact of previous refactoring, good catch again, thanks! Fixed.

stof commented on composer.json:
```diff
@@ -1,25 +1,26 @@
-    "description": "HTMLPurifier integration for your Symfony2 project",
-    "keywords": ["htmlpurifier"],
+    "description": "HTMLPurifier integration for your Symfony project",
+    "keywords": ["htmlpurifier", "html", "purifier", "symfony3", "symfony4"],
```
keyword should be `symfony`, not `symfony3` or `symfony4`

HeahDude replied:
Fixed.

stof commented on composer.json:
```diff
@@ -1,25 +1,26 @@
-        "twig/twig": "~1.3|~2.0"
+        "phpunit/phpunit": "^6.5.4",
+        "symfony/form": "~3.4.1 || ^4.0.1",
+        "twig/twig": "~1.6.5 || ~2.0"
```
runtimes are much newer than 1.6.5 IIRC. So this might be the wrong lowest bound.

HeahDude replied:
Runtime have been introduced in 1.26 https://github.com/twigphp/Twig/commit/9cb6860e5319544c37226e8a1a87f71eddaef5ab#diff-16e3dbccf13c8c447eff98d226c5d89bR813.
But this is more about namespaces, so I changed to 1.34|2.4. Thanks for the hint.

stof commented on composer.json:
```diff
@@ -1,25 +1,26 @@
-        "php": ">=5.3.2",
-        "symfony/framework-bundle": "~2.0|~3.0",
+        "php": "^5.5.9|>=7.0.8",
+        "symfony/framework-bundle": "~3.4.1 || ^4.0.1",
```
does it actually depend on FrameworkBundle ? Or only on some other components ?

HeahDude replied:
Hmm, good question. I guess the http kernel could be added with the others require-dev deps for the cache warmer, keeping only the DI in the required ones. WDYT?

HeahDude replied:
I've done that.

stof commented on composer.json:
```diff
@@ -1,25 +1,26 @@
     "extra": {
         "branch-alias": {
-            "dev-master": "1.0.x-dev"
+            "dev-master": "4.0.x-dev"
```
this change looks wrong to me. Why making it a 4.0 release ? This would be a 2.0 one at most

HeahDude replied:
A request from the maintainer in one of the first comment of this PR.

spolischook replied:
I prefer to be consistent with symfony version

stof replied:
but the bundle supports both 3.x and 4.x. So synchronized version number does not make sense. Thus, it means you restrict yourselves to do major versions every 2 years when Symfony does them, instead of doing them when *you* need them.

even official Symfony bundles developed in separate repos (MonologBundle, SwiftmailerBundle, MakerBundle) have separate version numbers (3.3.0, 3.2.2, 1.5.0 respectively).

HeahDude commented:
Thanks again @stof. Comments addressed.

javiereguiluz commented on DependencyInjection/Compiler/HTMLPurifierPass.php:
```diff
@@ -0,0 +1,56 @@
+
+        $purifiers = [];
+
+        foreach ($container->findTaggedServiceIds('exercise.html_purifier') as $id => $tags) {
```
Here (and in the error message) you could use the `PURIFIER_TAG` constant instead of the `exercise.html_purifier` string, right?

javiereguiluz commented on Tests/HTMLPurifiersRegistryTest.php:
```diff
@@ -0,0 +1,68 @@
+<?php
```
Blank line here after `<?php`

HeahDude commented:
@javiereguiluz All fixed, nice catch! Thank you for your review :)

HeahDude commented on DependencyInjection/Compiler/HTMLPurifierPass.php:
```diff
@@ -0,0 +1,56 @@
+        try {
+            $registry = $container->findDefinition(HTMLPurifiersRegistryInterface::class);
+        } catch (ServiceNotFoundException $e) {
+            return;
```
There won't be any definition and the check will fail, we need to check for an alias here.

stof commented on HTMLPurifiersRegistry.php:
```diff
@@ -0,0 +1,31 @@
+{
+    private $purifiersLocator;
+
+    public function __construct(ContainerInterface $purifiersLocator)
```
which doc would be needed here ? Only duplicating the typehint would be useless

HeahDude replied:
Not sure it would be valuable here.

XWB replied:
I would swear Symfony did something similar. Apparently not.

HeahDude commented on DependencyInjection/Compiler/HTMLPurifierPass.php:
```diff
@@ -0,0 +1,56 @@
+            $purifier = $container->getDefinition($id);
+
+            if (empty($purifier->getArguments())) {
+                $configId = "exercise_html_purifier.config.$profile";
```
This syntax may be used in the core, it mainly depends on the contextual readability it provides versus perf. This is more performant like this :)

HeahDude commented on Form/Listener/HTMLPurifierListener.php:
```diff
@@ -0,0 +1,56 @@
+        $this->profile = $profile;
+    }
+
+    public function purifySubmittedData(FormEvent $event)
```
Following Symfony style here, we don't need to add `: void`.

HeahDude commented on Form/TypeExtension/HTMLPurifierTextTypeExtension.php:
```diff
@@ -0,0 +1,63 @@
+{
+    private $purifiersRegistry;
+
+    public function __construct(HTMLPurifiersRegistryInterface $registry)
```
What would be the benefit?

HeahDude commented:
So should I make it a 2.0 before squash? ping @spolischook

HeahDude commented on Form/Listener/HTMLPurifierListener.php:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    /**
+     * @inheritdoc}
```
CS

XWB commented:
IMO it should be 2.0, this PR supports both Symfony 3 and 4 anyway.

spolischook commented:
@HeahDude as mentioned stof and XWB lets go with 2.0 version.
Thanks for your work!

HeahDude commented:
This is now ready to merge from my side. Once done I'll reopen https://github.com/heahprod/HTMLPurifierBundle/pull/2 on master in this repo.

HeahDude commented:
@spolischook rebased. Tests are green. Let's merged?

HeahDude commented:
@spolischook heahprod#2 is rebased with green tests, and ready to be open on master after merge before we can review, merge and release v2 :).

spolischook commented:
Thanks @HeahDude for your contribution

HeahDude commented:
Thank you for merging! Here's the following #52 :).
